ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -715,22 +757,70 @@ instance Arbitrary MempoolCapTestSetup where
+    -- just enough to hold the @nbInitialTxs@ initial transactions generated.
+    -- However, for the mempool capacity test, we don't want to prefill the
+    -- mempool with transactions but, rather, generate our own set of valid
+    -- transactions to manually add to the mempool. Because of this, we've set
+    -- @nbInitialTxs@ to 0 which will result in a generated mempool capacity
+    -- of 0 bytes.
+    --
+    -- Therefore, once we generate the valid transactions to use in the test,
+    -- we will adjust the mempool capacity before returning.
+    (testSetup, ledger) <- genTestSetup nbInitialTxs
 
+    -- Generate some valid transactions that will be added to the mempool as
+    -- part of a mempool capacity test.
     (vtxs, _) <- genValidTxs nbNewTxs ledger
-    pure MempoolCapTestSetup { mctsTestSetup = testSetup
+
+    let vTestGenTxs = map TestGenTx vtxs
+        txSizes = map txSize vTestGenTxs
+        -- The mempool should /at least/ be able to hold the largest of our
+        -- transactions. If we don't guarantee this, it's possible that a call
+        -- to 'addTxs' could block forever when trying to add a transaction
+        -- that has a size larger than the capacity. If the mempool capacity
+        -- is at least that of our largest transaction, this is no longer a
+        -- possibility.
+        capacityMinBound = maximum txSizes
+        -- The maximum mempool capacity should be the sum of all of the
+        -- transaction sizes.
+        capacityMaxBound = sum txSizes
+    capacity <- choose
+      ( capacityMinBound
+      , capacityMaxBound
+      )
+
+    -- As mentioned above, we need to  adjust the originally generated mempool
+    -- capacity with our newly generated one.
+    let testSetup' = testSetup { testMempoolCap = MempoolCapacity capacity }
+
+    pure MempoolCapTestSetup { mctsTestSetup = testSetup'
                              , mctsValidTxs  = vtxs
-                             , mctsCapacity  = MempoolCapacity (fromIntegral capacity)
+                             , mctsCapacity  = MempoolCapacity capacity
                              }
 
+-- | The 'MempoolSize' for a mempool consisting of no transactions.
+zeroMempoolSize :: MempoolSize
+zeroMempoolSize = MempoolSize { msNumTxs = 0, msNumBytes = 0 }
+
+-- | Split the given transactions at the point at which those on the left-hand
+-- side would be just enough to fill the mempool to capacity.
+splitTxsUntilCap :: [GenTx TestBlock]
+                 -> MempoolCapacity
+                 -> ([GenTx TestBlock], [GenTx TestBlock])
+splitTxsUntilCap txs (MempoolCapacity mpCap) = go 0 [] txs
+  where
+    go :: Word32
+        -> [GenTx TestBlock]
+        -> [GenTx TestBlock]
+        -> ([GenTx TestBlock], [GenTx TestBlock])
+    go _ accTxs [] = (accTxs, [])
+    go accByteSize accTxs (t:ts)
+      | (txSize t + accByteSize) <= mpCap =
+        go (accByteSize + txSize t) (accTxs ++ [t]) ts
```

Comment:
Actually, this is *O(n^2)*!

Instead of `accTxs ++ [t]`, which is *O(n)*, use `(t:accTxs)` (*O(1)*) and `reverse` (*O(n)*) the accumulator when you return it (on L818 and L822). I know this is just test code, its performance doesn't really matter in practice, and you'll update it in #1346, but still :slightly_smiling_face: